File: python/builder/HttpRequest.py
from HttpRequestConfig import HttpRequestConfig
from typing import Any
import requests
import json

class HttpRequest:
    def __init__(self, config: HttpRequestConfig):
        self._config = config

    def send(self) -> Any:
        attempt = 0

        while True:
            try:
                response = requests.request(
                    method=self._config.method,
                    url=self._config.url,
                    headers=self._config.headers,
                    data=json.dumps(self._config.body)
                    if self._config.body is not None
                    else None,
                    timeout=self._config.timeout,
                )

                if not response.ok:
                    # Retry solo si es error 5xx
                    if 500 <= response.status_code < 600 and attempt < self._config.retries:
                        attempt += 1
                        continue
                    raise Exception(f"HTTP error: {response.status_code}")

                return response.json()

            except requests.Timeout:
                if attempt < self._config.retries:
                    attempt += 1
                    continue
                raise Exception("Request timeout")

            except Exception:
                if attempt < self._config.retries:
                    attempt += 1
                    continue
                raise

File: ts/builder/HttpRequest.ts
import { HttpRequestConfig } from "./types.d";

export class HttpRequest {
  constructor(private readonly config: HttpRequestConfig) {}

  private async fetchWithTimeout(controller: AbortController): Promise<Response> {
    const timeoutId = setTimeout(() => {
      controller.abort();
    }, this.config.timeout);

    try {
      return await fetch(this.config.url, {
        method: this.config.method,
        headers: this.config.headers,
        body: this.config.body
          ? JSON.stringify(this.config.body)
          : undefined,
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timeoutId);
    }
  }

  public async send<T = unknown>(): Promise<T> {
    const maxRetries = this.config.retries ?? 0;
    let attempt = 0;

    while (true) {
      const controller = new AbortController();

      try {
        const response = await this.fetchWithTimeout(controller);

        if (!response.ok) {
          // Retry solo si es error 5xx
          if (response.status >= 500 && attempt < maxRetries) {
            attempt++;
            continue;
          }
          throw new Error(`HTTP error: ${response.status}`);
        }

        return response.json() as Promise<T>;
      } catch (error) {
        if (attempt < maxRetries) {
          attempt++;
          continue;
        }

        if (error instanceof DOMException && error.name === "AbortError") {
          throw new Error("Request timeout");
        }

        throw error;
      }
    }
  }
}
